proposals/0000-local-do.rst

```diff
@@ -0,0 +1,281 @@
+Local do
+========
+
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/216>`_.
+.. sectnum::
+.. contents::
+
+This proposal introduces a new extension ``-XLocalDo`` which makes it possible to overload the meaning of a do-notation expression *on a case-by-case basis* (as opposed to the global effect of ``-XRebindableSyntax``), by writing ``do @builder``. The design is inspired by F#'s  `computational expressions <https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions>`_.
+
+
+Motivation
+------------
+
+There are many kinds of monad-like things out there:
+
+* monads
+* indexed monads
+* graded monads
+* relative monads
+* linear variants of all the above once `linear types
+  <https://github.com/ghc-proposals/ghc-proposals/pull/111>`_ are
+  implemented
+  * in particular, the linear IO monad from the `Linear Haskell paper
+    <https://arxiv.org/abs/1710.09756>`_ is an example of a linear
+    graded monad
+* …
+
+All of these are theoretically compatible with the do-notation. And, in fact, really want the do-notation to work for them. After all, the entire existence of the do-notation can be ascribed to: “it's really annoying to program with ``(>>=)`` all the time”.
+
+The prescribed solution to use ``-XRebindableSyntax``. But ``-XRebindableSyntax`` is a very blunt instrument:
+
+* It affects many syntactic constructs (numerical literals, the ``if then else`` syntax, … (see the `full list <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-RebindableSyntax>`_)).
+* It implies ``-XNoImplicitPrelude``
+* It uses the same rebinding for all the do-expression in an entire file.
+
+You may not want all this. For instance, with linear types ``if then else`` syntax cannot be meaningfully rebound to a function. And there is no reason why a file wouldn't have a piece code referring to a monad, one to a graded monad, and one to a linear relative monad.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal creates a new language extension ``-XLocalDo``.
+
+When ``-XLocalDo`` is activated, the syntax of the ``do`` notation is changed to
+
+::
+
+  <lexp> ⟶ do [@<aexp>]
+
+(``aexp`` means that the notation after the ``@`` is parsed as a variable, unless there are parentheses)
+
+The additional expression is called the *builder* of the do-expression. A builder's type must be a record, other than that the type is immaterial.
```

Comment:
“Record type” isn't really a well-defined concept. Maybe drop that restriction? After all, I would not be completely surprised if `do @() e` would desugar to `e` without any problem.

Reply:
Fair enough.